javaslang/src/main/java/javaslang/collection/JavaConverters.java

```diff
@@ -0,0 +1,437 @@
+            return setDelegateAndCheckChanged(getDelegate().removeAll(that));
+        }
+
+        @Override
+        public boolean retainAll(Collection<?> collection) {
+            // may throw a ClassCastException accordingly to j.u.List.retainAll(Collection)
+            @SuppressWarnings("unchecked") final Collection<T> that = (Collection<T>) collection;
+            return setDelegateAndCheckChanged(getDelegate().retainAll(that));
+        }
+
+        @Override
+        public T set(int index, T element) {
+            // may throw an IndexOutOfBoundsException accordingly to the j.u.List.set(int, T)
+            return setDelegateAndGetPreviousElement(index, getDelegate().update(index, element));
+        }
+
+        @Override
+        public int size() {
+            return getDelegate().size();
+        }
+
+        @Override
+        public List<T> subList(int fromIndex, int toIndex) {
+            // may throw IndexOutOfBoundsException accordingly to j.u.List.subList(int, int)
+            return new ListView<>(getDelegate().subSequence(fromIndex, toIndex), isMutable());
+        }
+
+        @Override
+        public Object[] toArray() {
+            return getDelegate().toJavaArray();
+        }
+
+        @Override
+        public <T1> T1[] toArray(T1[] array) {
+            return getDelegate().toJavaList().toArray(array);
+        }
+
+        // -- Object.*
+
+        @Override
+        public boolean equals(Object o) {
+            if (o == this) {
+                return true;
+            } else if (o instanceof java.util.List) {
+                return Collections.areEqual(getDelegate(), (java.util.List<?>) o);
+            } else {
+                return false;
+            }
+        }
+
+        @Override
+        public int hashCode() {
+            return getDelegate().hashCode();
+        }
+
+        @Override
+        public String toString() {
+            return getDelegate().mkString("[", ", ", "]");
+        }
+
+        // -- private helpers
+
+        private boolean setDelegateAndCheckChanged(Seq<T> delegate) {
+            // TODO: or `return previousDelegate != delegate`?
+            return setDelegate(delegate).size() != delegate.size();
+        }
+
+        private T setDelegateAndGetPreviousElement(int index, Seq<T> delegate) {
+            return setDelegate(delegate).get(index);
+        }
+
+        private static class ListIterator<T> extends HasDelegate<IndexedSeq<T>> implements java.util.ListIterator<T> {
+
+            private static final long serialVersionUID = 1L;
+
+            private int index;
+
+            static <T> ListIterator<T> of(Seq<T> seq, int index) {
+                final IndexedSeq<T> indexedSeq = (seq instanceof IndexedSeq) ? (IndexedSeq<T>) seq : seq.toVector();
+                return new ListIterator<>(indexedSeq, index);
+            }
+
+            ListIterator(IndexedSeq<T> delegate, int index) {
+                super(delegate, true); // TODO: always mutable?
```

Comment:
@seanf

> You know what I'm going to say to that question. :-)

Yes :-)

Yep, you are right. In the immutable case we should throw. Here is a test:

```java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.ListIterator;

public class Test {

    public static void main(String[] args) {
        Test test = new Test();
        test.testMutable();
        test.testImmutable();
    }

    void testMutable() {
        List<Integer> list = new ArrayList<>();
        list.add(1);
        list.add(2);
        list.add(3);
        testListIterator(list);
    }

    void testImmutable() {
        List<Integer> list = Arrays.asList(1, 2, 3);
        testListIterator(list);
    }

    <T> void testListIterator(List<T> list) {
        ListIterator<T> iter = list.listIterator();
        iter.next();
        iter.remove();
        System.out.println(list);
    }
}
```

Output:

```bash
[2, 3]
Exception in thread "main" java.lang.UnsupportedOperationException
	at java.util.AbstractList.remove(AbstractList.java:161)
	at java.util.AbstractList$Itr.remove(AbstractList.java:374)
	at javaslang.Test.testListIterator(Test.java:37)
	at javaslang.Test.testImmutable(Test.java:31)
	at javaslang.Test.main(Test.java:18)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
```